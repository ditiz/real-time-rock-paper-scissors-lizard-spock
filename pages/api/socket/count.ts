import { getIo } from "@/utils/socket"
import { Server as NetServer } from "http"
import { Socket } from "net"
import { NextApiRequest, NextApiResponse } from "next"
import { Server as ServerIO } from "socket.io"

export type NextApiResponseServerIO = NextApiResponse & {
  socket: Socket & {
    server: NetServer & {
      io: ServerIO
    }
  }
}

export const config = {
  api: {
    bodyParser: false,
  },
}

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponseServerIO
) {
  let io = res.socket.server.io
  if (!io) {
    // adapt Next's net Server to http Server
    const httpServer: NetServer = res.socket.server
    io = getIo(httpServer)
  }

  let nb = (io.engine as unknown as Record<string, unknown>).clientsCount
  // append SocketIO server to Next.js socket server response
  res.socket.server.io = io
  res.status(200).json(nb)
}
