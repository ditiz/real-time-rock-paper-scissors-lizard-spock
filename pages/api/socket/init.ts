import { Server as NetServer } from "http";
import { Socket } from "net";
import { NextApiRequest, NextApiResponse } from "next";
import { Server as ServerIO } from "socket.io";

export type NextApiResponseServerIO = NextApiResponse & {
  socket: Socket & {
    server: NetServer & {
      io: ServerIO;
    };
  };
};

export const config = {
  api: {
    bodyParser: false,
  },
};

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponseServerIO
) {
  if (!res.socket.server.io) {
    console.info("New Socket.io server...");
    // adapt Next's net Server to http Server
    const httpServer: NetServer = res.socket.server;
    const io = new ServerIO(httpServer, {
      path: "/api/socket/init",
      cors: {
        origin: "*",
        methods: ["GET", "POST"],
      },
    });

    io.on("connection", (socket) => {
      io.emit("new-user", {
        latestConnection: socket.id,
        connectionCount: (io.engine as unknown as Record<string, unknown>)
          .clientsCount,
      });

      socket.on("user-action", (args) => {
        console.info("user-action", args.choice);
        io.emit("rcv-action", args.choice, socket.id);
      });

      socket.on("disconnect", () => {
        console.log("a user disconnected");
      });
    });

    // append SocketIO server to Next.js socket server response
    res.socket.server.io = io;
  }
  res.end();
}
