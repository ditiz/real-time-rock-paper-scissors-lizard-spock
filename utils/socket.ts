import { Namespace } from "@/types/namespace"
import { Server, Socket } from "socket.io"

export const socketGameClassic = (io: Server) => {
  const namespaceIo = io.of(Namespace.classic)
  namespaceIo.on("connection", (socket: Socket) => {
    namespaceIo.emit("new-user", {
      latestConnection: socket.id,
      connectionCount: (io.engine as unknown as Record<string, unknown>)
        .clientsCount,
    })

    socket.on("user-action", (args) => {
      console.info("user-action", args.choice)
      namespaceIo.emit("rcv-action", args.choice, socket.id)
    })

    socket.on("disconnect", () => {
      console.info("a user disconnected")
    })
  })
}

export const socketGameAdvanced = (io: Server) => {
  const namespaceIo = io.of(Namespace.advanced)
  namespaceIo.on("connection", (socket: Socket) => {
    namespaceIo.emit("new-user", {
      latestConnection: socket.id,
      connectionCount: (io.engine as unknown as Record<string, unknown>)
        .clientsCount,
    })

    socket.on("user-action", (args) => {
      console.info("user-action", args.choice)
      namespaceIo.emit("rcv-action", args.choice, socket.id)
    })

    socket.on("disconnect", () => {
      console.info("a user disconnected")
    })
  })
}
