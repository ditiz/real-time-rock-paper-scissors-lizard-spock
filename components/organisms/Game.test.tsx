import "@testing-library/jest-dom"
import { render, screen } from "@testing-library/react"
import { act } from "react-dom/test-utils"
import { Socket } from "socket.io"
import GameOrganism from "./Game"

it("renders GameOption unchanged", () => {
  const mockSocket = {
    on: jest.fn(),
    emit: jest.fn(),
  }

  const { container } = render(
    <GameOrganism socket={mockSocket as unknown as Socket} />
  )

  expect(container).toMatchSnapshot()
})

test("should listening for event 'rcv-action'", () => {
  const mockSocket = {
    on: jest.fn(),
    emit: jest.fn(),
  }

  render(<GameOrganism socket={mockSocket as unknown as Socket} />)
  expect(mockSocket.on).toBeCalledWith("rcv-action", expect.anything())
})

test("should emit event 'user-action'", () => {
  const mockSocket = {
    on: jest.fn(),
    emit: jest.fn(),
  }

  render(<GameOrganism socket={mockSocket as unknown as Socket} />)

  act(() => {
    screen.getAllByRole("img")[0].click()
  })

  expect(mockSocket.emit).toBeCalledWith("user-action", expect.anything())
})
