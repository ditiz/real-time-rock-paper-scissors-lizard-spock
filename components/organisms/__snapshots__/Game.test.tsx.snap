// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders GameOption unchanged 1`] = `
<div>
  <h2
    class="game__title"
  >
    Choose your move
  </h2>
  <article
    class="game"
  >
    <section>
      <img
        alt="rock icons"
        data-nimg="1"
        decoding="async"
        height="50"
        loading="lazy"
        src="icons/rock.svg"
        style="color: transparent;"
        width="50"
      />
    </section>
    <section>
      <img
        alt="paper icons"
        data-nimg="1"
        decoding="async"
        height="50"
        loading="lazy"
        src="icons/paper.svg"
        style="color: transparent;"
        width="50"
      />
    </section>
    <section>
      <img
        alt="scissors icons"
        data-nimg="1"
        decoding="async"
        height="50"
        loading="lazy"
        src="icons/scissors.svg"
        style="color: transparent;"
        width="50"
      />
    </section>
    <section>
      <img
        alt="lizard icons"
        data-nimg="1"
        decoding="async"
        height="50"
        loading="lazy"
        src="icons/lizard.svg"
        style="color: transparent;"
        width="50"
      />
    </section>
    <section>
      <img
        alt="spock icons"
        data-nimg="1"
        decoding="async"
        height="50"
        loading="lazy"
        src="icons/spock.svg"
        style="color: transparent;"
        width="50"
      />
    </section>
  </article>
</div>
`;
